### PR TITLE
feat: Add Controller

### PR DESCRIPTION
**Pull request by ntucker**
### Motivation

Usage has gone well, elevating https://resthooks.io/blog/2021/08/29/Experimental-useController as addition to core

### Solution

```ts
const controller = useController();

// actions
controller.fetch(UserResource.detail(), { id })
controller.receive(UserResource.detail(), { id }, response)
controller.receiveError(UserResource.detail(), { id }, error)
controller.invalidate(UserResource.detail(), { id })
controller.resetEntireCache()
controller.subscribe(UserResource.detail(), { id })
controller.unsubscribe(UserResource.detail(), { id })
```

fetch throttles all `sideEffect: undefined`. to force new fetch users should abort the existing first.

Future work

- [ ] Add controller to middlewareapi (for managers)
- [ ] Abortability - any fetch with endpoint, will get signal added `controller.abort(UserResource.detail(), { id })`
- [ ] Retrieval - replace useDenormalized
  ```ts
  // retrieval
  const state = useContext(StateContext);
  const [value, expiresAt] = controller.getResponse(state, UserResource.detail(), { id })
  const error = controller.getError(state, UserResource.detail(), { id })
  ```

### CODE REVIEW

File: packages/core/src/controller/Controller.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+  >(
+    endpoint: E,
+    ...args: readonly [...Parameters<E>]
+  ): ReturnType<E> => {
+    const action = createFetch(endpoint, {
+      args,
+    });
+    this.dispatch(action);
+
+    return action.meta.promise as any;
```
Comment (ljharb) [inline packages/core/src/controller/Controller.ts:53]: maybe?
```suggestion
    return action.meta.promise as Promise<ReturnType<E>>;
```

File: packages/rest-hooks/src/react-integration/provider/ExternalCacheProvider.tsx [hunk 2]
```diff
@@ -55,12 +57,22 @@ export default function ExternalCacheProvider<S>({
...
     <DispatchContext.Provider value={dispatch}>
       <StateContext.Provider value={optimisticState}>
-        <DenormalizeCacheContext.Provider value={denormalizeCache.current}>
-          {children}
-        </DenormalizeCacheContext.Provider>
+        <ControllerContext.Provider value={controller.current}>
+          <DenormalizeCacheContext.Provider
+            value={controller.current.globalCache}
+          >
+            {children}{' '}
```
Comment (ljharb) [inline packages/rest-hooks/src/react-integration/provider/ExternalCacheProvider.tsx:73]: the space seems very weird

**Review by ljharb — APPROVED** [review]

File: packages/rest-hooks/src/react-integration/provider/ExternalCacheProvider.tsx (hunk 2 above)
Comment (ntucker) [reply packages/rest-hooks/src/react-integration/provider/ExternalCacheProvider.tsx:73]: oh that was accident from formatting

File: packages/core/src/controller/Controller.ts (hunk 1 above)
Comment (ntucker) [reply packages/core/src/controller/Controller.ts:53]: ReturnType<E> is already a promise so i'll just put this in manually

**Pull request merged**